flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderRetrievalConnectionHandlingTest.java

```diff
@@ -0,0 +1,390 @@
+                            zooKeeperClient,
+                            retrievalPath,
+                            queueLeaderElectionListener,
+                            ZooKeeperLeaderRetrievalDriver.LeaderInformationClearancePolicy
+                                    .ON_SUSPENDED_CONNECTION,
+                            fatalErrorHandlerResource.getFatalErrorHandler());
+
+            writeLeaderInformationToZooKeeper(
+                    leaderRetrievalDriver.getConnectionInformationPath(),
+                    leaderAddress,
+                    UUID.randomUUID());
+
+            // do the testing
+            CompletableFuture<String> firstAddress = queueLeaderElectionListener.next();
+            assertThat(
+                    "The first result is expected to be the initially set leader address.",
+                    firstAddress.get(),
+                    is(leaderAddress));
+
+            closeTestServer();
+
+            CompletableFuture<String> secondAddress = queueLeaderElectionListener.next();
+            assertThat("The next result must not be missing.", secondAddress, is(notNullValue()));
+            assertThat(
+                    "The next result is expected to be null.",
+                    secondAddress.get(),
+                    is(nullValue()));
+        } finally {
+            if (leaderRetrievalDriver != null) {
+                leaderRetrievalDriver.close();
+            }
+        }
+    }
+
+    @Test
+    public void testSuspendedConnectionDoesNotClearLeaderInformationIfClearanceOnLostConnection()
+            throws Exception {
+        final String retrievalPath = "/testConnectionSuspendedHandling";
+        final String leaderAddress = "localhost";
+
+        final QueueLeaderElectionListener queueLeaderElectionListener =
+                new QueueLeaderElectionListener(1);
+        ZooKeeperLeaderRetrievalDriver leaderRetrievalDriver = null;
+        try {
+            leaderRetrievalDriver =
+                    new ZooKeeperLeaderRetrievalDriver(
+                            zooKeeperClient,
+                            retrievalPath,
+                            queueLeaderElectionListener,
+                            ZooKeeperLeaderRetrievalDriver.LeaderInformationClearancePolicy
+                                    .ON_LOST_CONNECTION,
+                            fatalErrorHandlerResource.getFatalErrorHandler());
+
+            writeLeaderInformationToZooKeeper(
+                    leaderRetrievalDriver.getConnectionInformationPath(),
+                    leaderAddress,
+                    UUID.randomUUID());
+
+            // do the testing
+            CompletableFuture<String> firstAddress = queueLeaderElectionListener.next();
+            assertThat(
+                    "The first result is expected to be the initially set leader address.",
+                    firstAddress.get(),
+                    is(leaderAddress));
+
+            closeTestServer();
```

Comment:
nit: I would prefer to call `testingServer.restart()` here (assuming it would reconnect after it is back up) to avoid the potential of actually being counted as a lost connection.
In practice, this is irrelevant of course since we only wait for 100ms...

Reply:
I agree that restarting would be better. Same point as above though.

Reply:
Ah, I did not check exactly which code was old. I assumed this part was at least modified. It's fine either way!